xtensa: cavs: support running test by west flash and twister with remote host

enjiamai opened the pull request:
There are some changes in this PR:
1. Add a dedicate west runner "intel_adsp" for intel_adsp_cavs boards.
2. Add a client-server base firmware loader instead of original SSH-base one.
3. Add a support for running multiple intel_adsp_cavs boards at same time. This feature rely on the hardware map support for serial-pty. I has already split into another PR: https://github.com/zephyrproject-rtos/zephyr/pull/45197) and try to make this PR clean.

Now we can run the test on cavs boards by west flash command, ex:
```
west build -b intel_adsp_cavs25 samples/hello_world
west flash --remote-host [remote hostname or ip addr]
```
And if you want to specify the location of SOF tool instead of the default ones, such as rimage tool, config and the key,
run it by:
```
west flash --remote-host [remote hostname or ip addr]
                 --rimage-path [path to rimage tool] \
                 --config-path [path to .toml config file] \
                 --key [path to signing key]
```

#Run the twister:
```
twister -p intel_adsp_cavs18 --device-testing
--device-serial-pty="$ZEPHYR_BASE/soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.x.x,-l" \
--west-flash="--remote-host=192.168.x.x, --run-twister"
-T samples/hello_world -vv"
```

nashif commented on boards/xtensa/intel_adsp_cavs25/board.cmake:
```diff
@@ -1,7 +1,6 @@
 
-board_set_flasher_ifnset(misc-flasher)
-board_finalize_runner_args(misc-flasher)
+set(SUPPORTED_EMU_PLATFORMS adsp)
```
but this is not an emulated platform.

enjiamai replied:
Thanks for your review, @nashif , got it. working on not to use emu.

chen-png replied:
this is not a simulator, but I think you can extend the current BinaryHandler, make it supports user to invoke their own custom script to load any expected binary files(elf, bin and so on) to run their tests, like what we are doing using west-flash for custom flash runner.

nashif commented on boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25.yaml:
```diff
@@ -1,6 +1,7 @@
 identifier: intel_adsp_cavs25
 name: cAVS 2.5 Audio DSP (converged Audio Voice and Speech)
 type: mcu
+simulation: xt-sim
```
and we are not a simulator.

enjiamai replied:
I try to add a type "adsp" instead of "mcu" for intel cavs board. By setting this to make twister run cavs platform by BinaryHandler.

nashif replied:
why? we do not want to end up with everyone defining their own type. This is not the intended use for this key. I think you are taking the wrong approach completely and fixing an issue in the wrong place. Please start a discussion first about what you are trying to do and what problem are you trying to solve.

enjiamai replied:
Thanks for the review, @nashif , let me have a simple explanation please.
Actually I want to run the CAVS testing in a by the BianryHandler, because in the DeviceHander the output by the script cannot be handled well in serial-pty mode. The BianryHandler can handle it well directly. (Base on our current script to get the cavs output)

Except adding a new type, I think there are some way to do this:
(1) Add a extra parameter to twister, which will run the specified script by BinaryHandler. (ex --binary-runner to specify the cavstwist.sh and its arguments)
(2) Add another mode in the DeviceHandler besides serial and serial-pty mode, which behaves like the BinaryHandler. This willf also be benifit if we want to add a hardware-map for it.
(3) Change some behavior of cavstwister.sh and cavstool.py to fit the serial-pty, or even add a new runner for it (Like the intel_s1000_crb did)

So I think the critical problem for me is should we only run the CAVS boards by DeviceHandler only in twister? Thanks.

enjiamai replied:
I correct the wrong usage of it in the first version. The second version still use BinaryHandler to run. And this is the third version, it runs by the DeviceHandler and support without run with -j 1 well. Also the hardware map file can be used to specify the remote-host.  But one point is I make some extra change in device-serial-pty, to make it judge the output from the stdout. Could you please help to give some feedback?

nashif commented on cmake/emu/adsp.cmake:
```diff
@@ -0,0 +1,36 @@
+# SPDX-License-Identifier: Apache-2.0
```
this is mis-using the emulation support for actually loading images into actual hardware. We should not do this.

enjiamai replied:
Thanks. fixed it.

chen-png commented on boards/xtensa/intel_adsp_cavs25/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+  rimage
+  )
+
+if(CONFIG_BOARD_INTEL_ADSP_CAVS25 OR CONFIG_BOARD_INTEL_ADSP_CAVS25_TGPH)
```
this is a common step for all intel adsp platforms, I think we can put it in "soc/xtensa/intel_adsp/common/CMakeLists.txt".

enjiamai replied:
yes, good point, will try to do it this way :), thanks!

enjiamai replied:
Yes, I tried to put some common part in to "soc/xtensa/intel_adsp/common/CMakeLists.txt". For the "key" using to signed rimage I still put them in their own board.cmake.

mbolivar-nordic commented on soc/xtensa/intel_adsp/tools/cavstwist.sh:
```diff
@@ -4,122 +4,93 @@
 
 ########################################################################
 
 CAVSTOOL=$ZEPHYR_BASE/soc/xtensa/intel_adsp/tools/cavstool.py
```
So this is a west runner, that calls a shell script, that calls a python file?

This is kind of a mess in my opinion.

enjiamai replied:
Thanks for your review, @mbolivar-nordic . Would you mean it's better to use python script instead of shell script here?

mbolivar-nordic replied:
What I mean is that the extra indirection looks like suspect design to me. It looks like cavstool.py was added by @andyross in f581d59821f34ad2c1f320dc3597a0883e7c6792, and then cavstwist.sh was added by the same person in e71b1a1148c7ac176a56a1800162e954cb9d8eae specifically as a twister integration script, and now there is a west runner (python) that's calling  the twister integration script (shell) that's calling cavstool.py (python). That doesn't make much sense.

enjiamai replied:
Hi @mbolivar-nordic , I tried to make a dedicate west runner for it. It will call the cavstool.py (firmware loader and runner in the DUT host side). Could you please help to see if this implementation make sense or not, when you got a change, thank you!

mbolivar-nordic reviewed (commented):
No objections, but it looks like the whole area could use some clean-up.

enjiamai committed:
scripts: runner: add a intel_adsp west runner

Add a intel_adsp west runner instead of using misc-flasher for
intel_adsp boards. Now running the test on intel_adsp_cavs by:

  west build -b intel_adsp cavs25 ...
  west flash --remote-host [remote hostname or ip addr] \
             --rimage-tool [path to rimage tool] \
             --config-dir [path to dir of .toml config file] \
             --key [path to signing key]

The intel_adsp west depends on the cavstool_server.py running
on the remote host machine.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
soc: intel_adsp: add a client-server based cavstool tool

The cavstool_client.py is a client use to communicate with
the firmware loading and running server which dealing with
the requests from the intel_adsp west runner. It supports:

1. Download firmware to remote ADSP host and running.
2. Send the ADSP log messages back to client.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+        parser.add_argument('--remote-host',
+                            help='hostname of the remote targeting ADSP board')
+        parser.add_argument('-t', '--timeout', type=int, default=60,
+                            help='the pty where stdout goes to')
```
This help text looks incorrect.

enjiamai replied:
Fixed. Sorry for the mistake.

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+                self.flash(**kwargs)
+            else:
+                self.run_sim(**kwargs)
+        else:
+            self.logger.info('Only support flash cmd now. Do nothing...')
```
The `if` / `else` blocks are not necessary. The parent class will fail in `run()` before calling this `do_run()` method if the command is not supported in the `capabilities()` return value.

```python
        caps = self.capabilities()
        if command not in caps.commands:
            raise ValueError('runner {} does not implement command {}'.format(
                self.name(), command))
        self.do_run(command, **kwargs)
```

enjiamai replied:
Good to know, thanks! Remove it.

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+
+        try:
+            self.check_call(download_cmd)
+        except subprocess.CalledProcessError:
+            return 1
```
Please do not swallow exceptions like this. Let the exception propagate up; west should handle the CalledProcessError for you if this happens.

enjiamai replied:
Got it. I remove the try/except and let it being handled by west. Please kindly let me know if it is correct way to it, thank you!

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+        self.run_cmd = (['ssh', f'{self.remote_host}', 'sudo', './cavstool.py', 'zephyr.ri', '-t', f'{self.timeout}'])
+        self.kill_cmd = (['ssh', f'{self.remote_host}', 'sudo', 'pkill' , '-9', '-f', 'cavstool'])
+
+        self.logger.info(f"rcmd: {self.run_cmd}")
```
This sort of thing is usually done using `self.logger.debug` to avoid verbosity.

enjiamai replied:
Yes, corrected.

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+
+        try:
+            self.check_call(self.run_cmd)
+        except Exception as e:
+            self.logger.error(e)
+            self.do_cleanup()
+            return 1
```
Please don't swallow exceptions like this.

enjiamai replied:
Thanks for correcting this. Fix it.

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+            return 1
+
+        self.do_cleanup()
+        return 0
```
The return value is not used in the caller, so this is unecessary.

enjiamai replied:
Ok, fixed. Thanks.

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+
+    def do_cleanup(self):
+        self.check_call(self.kill_cmd)
+        self.logger.error(f'Kill the cavs FW loading tool in {self.remote_host}')
```
`error` looks incorrect here. This action is always done. Errors should only be logged when something has gone wrong.

enjiamai replied:
Remove this line due to the mechanism for stop the firmware tool is change. But I will remember your comments here, you are right, thanks.

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+            self.check_call(sim_run_cmd)
+        except Exception as e:
+            self.logger.error(e)
+            return 1
+
+        return 0
```
Same issues around swallowing exceptions and unused return values are here.

enjiamai replied:
Fixed, thanks!

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+
+        self.logger.info(f"rcmd: {self.run_cmd}")
+
+        # We need this to cleanup the remote host process.
+        signal.signal(signal.SIGTERM, self.do_cleanup)
```
Whoa, catching SIGTERM like this right here is very strange and looks suspicious. Why not just use `try: / finally:` below instead? If you really need to catch SIGTERM, please explain why.

enjiamai replied:
You totally got the point! I am aware of this is not a good usage. And after changing the way of stop the remote host service, I don't have to catch the INT signal to stop the remote cavstool.py when the process terminated any more. Remove it.

mbolivar-nordic reviewed (changes requested):
@enjiamai I want to thank you for taking the time to add a runner, especially since I was not blocking the pull request before. This looks a lot nicer now from the west side.

I have a few comments and change requests related to the runner and nearby commits. I will leave the rest up to you all at Intel.

mbolivar-nordic commented on soc/xtensa/intel_adsp/common/CMakeLists.txt:
```diff
@@ -100,3 +100,103 @@ add_custom_target(
+if(${RIMAGE} STREQUAL RIMAGE-NOTFOUND
+  OR ${RIMAGE_CONFIG_SCHEMA} STREQUAL RIMAGE_CONFIG_SCHEMA-NOTFOUND)
+
+  find_program(RIMAGE_BUILD_SCRIPT build_rimage.sh
+    PATH ${ZEPHYR_BASE}/soc/xtensa/intel_adsp/tools/)
+    message(WARNING "Start to build rimage tool by ${ZEPHYR_BASE}")
+
+  execute_process(
+    COMMAND
+    ${RIMAGE_BUILD_SCRIPT}
+    ${ZEPHYR_BASE}
+    ${ZEPHYR_SOF_MODULE_DIR}
+    RESULT_VARIABLE return_code
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+    WORKING_DIRECTORY ${ZEPHYR_SOF_MODULE_DIR}/rimage
+  )
+  if(return_code)
+    message(WARNING "build rimage tool failed: ${stderr}")
+  endif()
```
@stephanosio @nashif while @tejlmand is on leave, I won't block over this, but this commit looks a bit off to me.

It seems like ExternalProject_add() or something like that should be used instead if you really need this. It's not clear to me why the zephyr build system should be responsible for compiling a host tool, which is what I think rimage is.

stephanosio replied:
I agree with @mbolivar-nordic.

If `rimage` (a host tool) is required, downloading or building it should be an independent process from building the Zephyr image.

Please either provide the instruction for download/building the `rimage` in the documentation, or provide a west extension that does this (similar to how Espressif handles their toolchain installation).

In the long term, it may be preferable to add this host tool as part of the Zephyr SDK (see https://github.com/zephyrproject-rtos/sdk-ng/issues/462).

enjiamai replied:
Hi @mbolivar-nordic @stephanosio thanks for letting me know the right way to do it. I totally remove the "rimage tool build process", and specify the rimage build documentation link, as @nashif suggested. And I am also looking forward to https://github.com/zephyrproject-rtos/sdk-ng/issues/462. Thanks for the review.

enjiamai commented:
Hi @mbolivar-nordic , thanks so much for your review and I am working on addressing you comment :)

nashif reviewed (changes requested):
partial review, but the rimage comments need to be addressed. See https://github.com/zephyrproject-rtos/sdk-ng/issues/462 which should be the right approach for doing this.

enjiamai committed:
boards: xtensa: Activate the intel_adsp west runner

Make the intel_adsp west runner starting to work on all the
intel_adsp boards. Changes include:

1. Make the cavstool.py work as a service in remote host
   ADSP board and rename it to cavstool_server.py.

2. Active the runner and adds a common board.cmake file to
   specify the default signing key for cavs boards.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
soc: cavs: add a README to document the usage

Add a usage of how to run the intel_adsp_cavs board, when the
new intel_adsp west runner introduced.

Also update the docs of up_squared and TGL.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai commented on boards/common/adsp.board.cmake:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
```
Fixed. Thanks for correcting me this.

enjiamai commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+        self.timeout = timeout
+        self.bin_fw = os.path.join(cfg.build_dir, 'zephyr', 'zephyr.ri')
+
+        self.zephyr_base = os.environ.get('ZEPHYR_BASE')
```
Great! Your correction help me solve this problem.

enjiamai commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,118 @@
+        self.zephyr_base = os.environ.get('ZEPHYR_BASE')
+        self.cavstool = os.path.join(self.zephyr_base, DEFAULT_CAVSTOOL)
+        self.build_dir = cfg.build_dir
+        self.board_dir = cfg.board_dir
```
Good to know! Remove it and use the self.cfg.build_dir directly. Thanks!

enjiamai commented on soc/xtensa/intel_adsp/tools/build_rimage.sh:
```diff
@@ -0,0 +1,136 @@
+#!/bin/bash
```
Thanks for the review. I remove this script. And I change it to:  the zephyr.ri can be signed both in "build" or "west flash" stage both, depends on if you have rimage or singing key and toml config files in place.

enjiamai commented on soc/xtensa/intel_adsp/common/CMakeLists.txt:
```diff
@@ -100,3 +100,103 @@ add_custom_target(
+#     -DRIMAGE_KEY=path_to/xxx.key -DRIMAGE_CONFIG=path_to/xxx.toml
+#
+find_program(RIMAGE rimage
+  PATH ${ZEPHYR_SOF_MODULE_DIR}/rimage/build ${ZEPHYR_BASE})
```
Fixed it. thanks!

enjiamai commented on soc/xtensa/intel_adsp/common/CMakeLists.txt:
```diff
@@ -100,3 +100,103 @@ add_custom_target(
+find_file(RIMAGE_CONFIG_SCHEMA ${RIMAGE_TARGET}.toml
+  PATH ${ZEPHYR_SOF_MODULE_DIR}/rimage/config)
+
+# Try to build the rimage tool and get the toml config.
```
Agree! Fix it. thanks!

enjiamai commented:
I appreciate the review and I got lots of knowledge of west runner!  Hi, @mbolivar-nordic , @nashif , @MaureenHelm , @stephanosio, I have address all the comments. Could you please help with reviewing it again when you got a chance, thank you!

MaureenHelm commented on soc/xtensa/intel_adsp/tools/README:
```diff
@@ -0,0 +1,116 @@
+This documentation descripts how to run the intel_adsp_cavs board.
```
Please move this into rst and make sure it's included somewhere in the generated documentation.

enjiamai replied:
I updated it to the rst file and move this to boards/xtensa/intel_adsp_cavs15/doc/generic_intel_adsp.rst. Thanks for the review.

enjiamai commented:
Dear all, could you please help to review this PR again when you got a chance, and the major changes compare to previous PR are:
1. Try to split the twister hardware map support with the west runner itself.
2. Remove the signing image process in cmake I made previously to keep it simple. (If you suppose we can have it, then I can add it back anytime)
3. Add a rst document and also update the docs of cavs15 and cavs25.

Thanks for your review and all the comments!

MaureenHelm commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,131 @@
+        self.rimage_tool = rimage_tool
+        self.config_dir = config_dir
+        self.bin_fw = os.path.join(cfg.build_dir, 'zephyr', 'zephyr.ri')
+        self.bin_file = os.path.join(cfg.build_dir, 'zephyr', 'zephyr.bin')
```
I think you can just use `cfg.bin_file` instead of reconstructing the path

enjiamai replied:
Thanks for correcting this! And also I found it doesn't need bin_file now, so remove it directly. If there is in a case needing bin_fw again, I will use the cfg.bin_file directly, thanks.

nashif commented on boards/xtensa/intel_adsp_cavs15/doc/index.rst:
```diff
@@ -72,55 +72,80 @@ Loading image to Audio DSP
+The only way to debug application is using logging. Logging and ADSP logging
+backend needs to be enabled in the application configuration.
+
+ADSP logging backend writes logs to the ring buffer in the shared memory.
```
remove this, I do not think this is relevant, there are other way for getting log messages which will need to be supported as well.

enjiamai replied:
Yes, also removed.

nashif commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -0,0 +1,165 @@
+Intel Adsp Generic Running Guide
+################################
+
+This documentation descripts how to run the intel_adsp_cavs boards. Including:
```
describes...

nashif commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -0,0 +1,165 @@
+Now you can see the outout log in your terminal.
+
+
+If you don't want to use the default location of SOF tools, you can
```
I would talk about rimage directly, not SOF tools.

nashif commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -0,0 +1,130 @@
+
+        # When run by twister, it will not output the log
+        # message directly because the server's output log
+        # will grab by twister's device-serial-pty.
```
This looks like a hack, we should not have special handling for twister, please consider a more generic solution or name this option differently.

enjiamai replied:
Yes, but I have some dilemma here. I would like to make "west flash"  here not only flash but also show the output message in directly STDOUT. This way is convenient I think,  though this behavior is not as the same as the devices which use the physical serial port. (west flash just flash without output log).  It needs a way to tell it runs in twister or not. Your suggestion is good, I propose to rename it to "--pty" because this is related the where the log output.

And I made another change there, the --pty generate by twister then pass to west flash. By this we just reduce the parameter to one, ex:   --west-flash="--remote-host=192.168.1.x".  How do you think?

enjiamai replied:
I found a problem here but might be solved a little later. So I still keep it --west-flash="--remote-host=192.168.1.x,--pty" here in twister. Thanks!

enjiamai committed:
twister: passing a pty pararmeter to intel_adsp runner

Add a pty parameter from twister to intel_adsp when west-flash,
to tell the intel_adsp runner to change the way it outputs the
log.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai commented on boards/xtensa/intel_adsp_cavs15/doc/index.rst:
```diff
@@ -72,55 +72,80 @@ Loading image to Audio DSP
-udev trick the operator prefers.
+.. code-block:: console
+
+   west flash --remote-host cavs15
```
No, if we have: 
(1) rimage tool install in system path (such as: /usr/bin or /usr/local/bin ..) 
(2) config file:  If we follow the SOF's setup steps this will be in modules/audio/sof/rimage/config
(3) key:  the default key location will be: modules/audio/sof/keys.
Then west flash --remote-host will be only parameter needed.

enjiamai commented on boards/xtensa/intel_adsp_cavs15/doc/index.rst:
```diff
@@ -72,55 +72,80 @@ Loading image to Audio DSP
 
-Debugging
-=========
+It's a daemon that is in change of loading the firmware to ADSP, and send out the
```
It reads much better, thanks for the help!

enjiamai commented on boards/xtensa/intel_adsp_cavs15/doc/index.rst:
```diff
@@ -72,55 +72,80 @@ Loading image to Audio DSP
+ip address), and the user account is ``user``. Then copy the python tool to the
+``up_squared`` board from your building evironment:
+
+   $scp boards/xtensa/intel_adsp_cavs15/tools/cavstool_server.py user@cavs15:
```
Fixed, thanks!

enjiamai commented on soc/xtensa/intel_adsp/tools/cavstool_server.py:
```diff
@@ -595,29 +611,155 @@ async def main():
         if dsp.HIPCIDA & 0x80000000:
             dsp.HIPCIDA = 1<<31 # must ACK any DONE interrupts that arrive!
 
+class AdspRequestHandler(socketserver.BaseRequestHandler):
```
Got it fixed both in cavstool_client.py and cavstool_server.py. Thanks for the correction.

enjiamai commented on soc/xtensa/intel_adsp/tools/cavstool_server.py:
```diff
@@ -626,11 +768,36 @@ async def main():
+else:
+    HOST = get_host_ip()
+
 if __name__ == "__main__":
```
I think I fixed this exception message but it is not good enough. When I use Crtl-c t to stop it, The "port" resource that cavstool_server.py occupied will not be fully released.  That means it need to wait for about 30 sec to restart the cavstool_server.py. I am trying to better solving this.

enjiamai replied:
Got it fixed. Now it won't show the exception and will not report error even re-start right away.

enjiamai commented:
Excuse me @nashif , @MaureenHelm, @mbolivar-nordic  and all , could you please help to review it again when you got a chance, thank you!

enjiamai commented on scripts/west_commands/tests/test_imports.py:
```diff
@@ -24,6 +24,7 @@ def test_runner_imports():
                     'gd32isp',
                     'hifive1',
                     'intel_cyclonev',
+                    'intel_adsp',
```
Got it fixed both in test_import.py and __init__.py ! Thanks for the review.

enjiamai commented:
Rebase and correct alphabetical order only, thank for your review again.

mbolivar-nordic reviewed (approved):
Overall looks good to me. I have a few nitpicks about the runner implementation but I think it's a big improvement and it would be better to have it merged, so +1 from me. I guess this must wait for v3.2 though.

enjiamai commented:
> Overall looks good to me. I have a few nitpicks about the runner implementation but I think it's a big improvement and it would be better to have it merged, so +1 from me. I guess this must wait for v3.2 though.

No nitpick at all, @mbolivar-nordic , I appreciate you and other reviewers spending your time to review this and I got lots of benefit of understanding the runner, thank you so much! :)

kv2019i commented:
@enjiamai @aborisovich @marc-hb This unfortunately breaks the SOF CI currently. We've been relying on the cavstool.py and this PR now removes the old tool and replaces it with a client-server model. I'll talk with SOF CI folks how we can cope with this.

enjiamai commented:
Excuse me, @kv2019i , yes, it will change the way we run the cavstool, please let me know if I can do anything to mitigate it. or you need any message to set the new thing up. Thank you!

enjiamai commented:
Hi @kv2019i , could you please provide me some information about how you use the original cavstool.py in your CI ? Maybe I could help to speed up the migration, thanks!

kv2019i commented:
@enjiamai Our CI has used "cavstool.py -l" to gather logs and run the tool on the DUTs. We also have some deployment scripts that have assumed cavstool.py to exist in Zephyr tree, like in here https://github.com/thesofproject/sof/blob/main/scripts/xtensa-build-zephyr.py 

I'll let our CI folks to comment in, whether we should move to the client-server model, or should we try to restore the "local cavstool.py" variant of the tool.

enjiamai commented:
Thanks , @kv2019i , maybe restore the "local cavstool.py"  is a quick short-term solution. Beside, I am trying to submit a fix for it in https://github.com/zephyrproject-rtos/zephyr/pull/46308 , please kindly to review and comment it. if there is anything also need to be added. Sorry for the inconvenience.

marc-hb commented:
In the future can someone please make sure at least one SOF person is copied on backward incompatible changes like this before they're merged?

I'm not aware of any other, open-source ADSP user besides SOF, is there?

> could you please provide me some information about how you use the original cavstool.py in your CI ? 

A bit too late now but you can very simply submit any Zephyr PR to SOF CI with a one-line SOF PR: https://github.com/thesofproject/sof/blob/ebc222dab5686/scripts/xtensa-build-zephyr.py#L304

marc-hb commented:
Thanks Robin for drawing my attention to the `MAINTAINERS.yml` file: it seems to cover production code but not test code... typical? I will try to fix it.

marc-hb commented:
> A bit too late now but you can very simply submit any Zephyr PR to SOF CI with a ~one~ two-lines SOF PR

Real-world example where the fix for this is getting tested in SOF CI:
- https://github.com/thesofproject/sof/pull/5895

Anyone ever working on changes affecting SOF please have a look. Don't postpone, click now; it's literally 2 lines long and lets anyone run any unmerged Zephyr PR on SOF hardware. If you look at these 2 lines you will remember them, if not then you won't.

marc-hb commented:
> I think that means when there are changes related to SOF/CAVS, then we can get the SOF CI test result in advance. right?

Not automatically but manually yes, that's the whole point of this test "feature"/hack. Please use it!

Automatically would be great but there are a number of challenges, for a start SOF CI does not have the bandwidth to test every single Zephyr PR and that would not make sense anyway. Other challenges:
- https://github.com/thesofproject/sof/issues/3517
- https://github.com/zephyrproject-rtos/zephyr/discussions/34713

enjiamai commented:
I think it's quite nice already if we can manually submit then test it like this! will do it next time. :)

marc-hb commented:
> Thanks Robin for drawing my attention to the MAINTAINERS.yml file: it seems to cover production code but not test code...

I was wrong, the `MAINTAINERS.yml` file should have added SOF people, see below.

@nashif, @cfriedt  there seems to some `get_maintainer.py`-related glitch somewhere. Where is the `get_maintainer.py`invocation that assigns reviewers and corresponding log?
```
./scripts/get_maintainer.py commits e4c171ed2255~5..e4c171ed2255
Intel Platforms (Xtensa)
	status: maintained
	maintainers: nashif
	collaborators: andyross, dcpleung, lyakh, lgirdwood, mengxianglinx, marc-hb, kv2019i, MaureenHelm
	inform: 
	labels: platform: Intel CAVS
	description: 

Twister
	status: maintained
	maintainers: nashif
	collaborators: chen-png, galak, PerMac, enjiamai, hakehuang, gopiotr
	inform: 
	labels: area: Twister
	description: 

West
	status: maintained
	maintainers: mbolivar-nordic
	collaborators: carlescufi, swinslow
	inform: 
	labels: area: West
	description: 
```

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool_server.py:
```diff
@@ -11,6 +11,22 @@
 import argparse
+import socketserver
+import threading
+import netifaces
```
I wonder if `netifaces` shouldn't be a part of `requirements-run-test.txt` or so.

enjiamai replied:
You concern is right. I will try to add it to the requirements-run-test.txt. thanks! @edersondisouza

enjiamai replied:
Submit https://github.com/zephyrproject-rtos/zephyr/pull/46552 for it.
